velox/functions/lib/string/StringImpl.h

```diff
@@ -688,4 +689,115 @@ FOLLY_ALWAYS_INLINE void pad(
       padPrefixByteLength);
 }
 
+namespace detail {
+
+template <bool strictSpace>
+inline bool isSpaceAscii(unsigned char ch) {
+  if constexpr (strictSpace) {
+    return ch == ' ';
+  } else {
+    return std::isspace(ch);
+  }
+}
+
+template <bool strictSpace>
+inline bool isSpaceUtf8(utf8proc_int32_t cp) {
+  if constexpr (strictSpace) {
+    return cp == 0x20;
+  } else {
+    return isUnicodeWhiteSpace(cp);
+  }
+}
+
+template <bool strictSpace, typename TOutString, typename TInString>
+FOLLY_ALWAYS_INLINE void initcapAsciiImpl(
+    TOutString& output,
+    const TInString& input) {
+  output.resize(input.size());
+  const char* inputChars = input.data();
+  char* outputChars = output.data();
+
+  bool isStartOfWord = true;
+  for (size_t i = 0; i < input.size(); ++i) {
+    unsigned char currentChar = static_cast<unsigned char>(inputChars[i]);
+
+    if (isSpaceAscii<strictSpace>(currentChar)) {
+      isStartOfWord = true;
+      outputChars[i] = currentChar;
+    } else if (isStartOfWord) {
+      outputChars[i] = std::toupper(currentChar);
+      isStartOfWord = false;
+    } else {
+      outputChars[i] = std::tolower(currentChar);
+    }
+  }
+}
+
+template <bool strictSpace, typename TOutString, typename TInString>
+FOLLY_ALWAYS_INLINE bool initcapUtf8Impl(
+    TOutString& output,
+    const TInString& input) {
+  const char* inputBytes = input.data();
+  const char* inputEnd = inputBytes + input.size();
+
+  output.resize(input.size() * 4);
+  char* outputStart = output.data();
+  char* outputBytes = outputStart;
+
+  bool isStartOfWord = true;
+
+  while (inputBytes < inputEnd) {
+    utf8proc_int32_t originalCodepoint;
+    auto numBytesRead = utf8proc_iterate(
+        reinterpret_cast<const uint8_t*>(inputBytes),
+        inputEnd - inputBytes,
+        &originalCodepoint);
+    if (numBytesRead < 0) {
+      return false;
+    }
+
+    if (isSpaceUtf8<strictSpace>(originalCodepoint)) {
+      isStartOfWord = true;
+      // Copy delimiter as is.
+      std::memcpy(outputBytes, inputBytes, numBytesRead);
+      outputBytes += numBytesRead;
+    } else if (isStartOfWord) {
+      auto upperSize = upperUnicode(
+          outputBytes,
+          static_cast<size_t>(outputStart + output.size() - outputBytes),
+          inputBytes,
+          numBytesRead);
+      outputBytes += upperSize;
+      isStartOfWord = false;
+    } else {
+      auto lowerSize = lowerUnicode<strictSpace>(
```

Comment:
This might look a bit strange to pass `strictSpace` as `forSpark`. I'd like to rename the `forSpark` parameter of `lowerUnicode` function and make a refactor after this PR.

Reply:
I'll add a template parameter `turkishCasing` for the initcap function in https://github.com/facebookincubator/velox/pull/13993 after this PR got landed.